src/nupic/research/connections.py

```diff
@@ -52,92 +146,164 @@ def __init__(self,
 
     # Save member variables
     self.numCells = numCells
+    assert maxSegmentsPerCell > 0
+    assert maxSynapsesPerSegment > 0
     self.maxSegmentsPerCell = maxSegmentsPerCell
     self.maxSynapsesPerSegment = maxSynapsesPerSegment
 
-    # Mappings
-    self._segments = dict()
-    self._synapses = dict()
+    self._cells = [CellData() for _ in xrange(numCells)]
+    self._synapsesForPresynapticCell = defaultdict(list)
+    self._segmentForFlatIdx = []
 
-    # Indexes into the mappings (for performance)
-    self._segmentsForCell = dict()
-    self._synapsesForSegment = dict()
-    self._synapsesForPresynapticCell = defaultdict(dict)
+    self._numSegments = 0
+    self._numSynapses = 0
+    self._nextFlatIdx = 0
+    self._iteration = 0
 
-    # Index of the next segment to be created
-    self._nextSegmentIdx = 0
-    # Index of the next synapse to be created
-    self._nextSynapseIdx = 0
 
+  def segmentsForCell(self, cell):
+    """ Returns the segments that belong to a cell.
 
-  def cellForSegment(self, segment):
-    """ Returns the cell that a segment belongs to.
+    @param cell (int) Cell index
+
+    @return (set) Segment indices
+    """
+
+    segmentsData = self._cells[cell].segments
+    segments = []
+    for i in xrange(len(segmentsData)):
+      if not segmentsData[i].destroyed:
+        segments.append(Segment(i, cell))
+
+    return segments
+
+
+  def synapsesForSegment(self, segment):
+    """ Returns the synapses on a segment.
 
     @param segment (int) Segment index
 
-    @return (int) Cell index
+    @return (set) Synapse indices
     """
-    return self._segments[segment]
+    segmentData = self.dataForSegment(segment)
+    synapses = []
+    if segmentData.destroyed:
+      raise ValueError("Attempting to access destroyed segment's synapses")
 
+    for i in xrange(len(segmentData.synapses)):
+      if not segmentData.synapses[i].destroyed:
+        synapses.append(Synapse(i, segment))
 
-  def columnForSegment(self, segment, cellsPerColumn):
-    """ Returns the column that a segment's presynapticCell belongs to
+    return synapses
 
-    @param segment        (int) Segment index
-    @param cellsPerColumn (int) Number of cells in a column in the tm
 
-    @return (int)
+  def dataForSynapse(self, synapse):
+    """ Returns the data for a synapse.
+
+    @param synapse (Object) Synapse object
+
+    @return (SynapseData) Synapse data
     """
-    return self._segments[segment] / cellsPerColumn
 
+    return self._cells[synapse.segment.cell].segments\
+                      [synapse.segment.idx].synapses[synapse.idx]
```

Comment:
I'm not sure if it will make a perf difference, but definitely feel free to do something like:

~~~python
    segment = synapse.segment

    return self._cells[segment.cell].segments\
                      [segment.idx].synapses[synapse.idx]
~~~

for this and for `dataForSegment` to avoid lookups. I didn't do this in the C++ because there's no overhead in C to do `foo.bar.baz` lookups, since it's all just a single compile-time binary offset no matter how long the chain is.

Your call.